### PR TITLE
Refactored Docker environment.

### LINKED ISSUE #16642: Redesign docker environment
#### Issue Description

[docker-compose.yml](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/docker-compose.yml) is outdated, leading to broken installation of the project.

As discussed with @systimotic in Gitter room, I'll build relevant development environment for the project. 

## Tasks to do:

- [x] Rename `docker-compose.yml` to `sample.docker-compose.yml` and add` docker-compose.yml` to `.gitignore` - this way developers will be able to copy sample file and create their own `docker-compose.yml` files, giving them flexibility to make changes as they need.

### Define docker-compose services:
- [x] database service - mongo
- [x] web server - node
- [x] mail system - mailhog 

### Installation process:
- [x] Follow the project installation steps and make sure it leads to working instance of the project inside Docker environment. 
- [x] Document everything under project's installation section.

### PR DESCRIPTION
**Pull request by otarza**
docker-compose.yml renamed to sample.docker-compose.yml.
docker-compose.yml added to .gitignore so every developer can copy it from sample.docker-compose.yml and make changes according to their local needs.

docs(Docker): Updated docs for Docker install.
Added Docker specific installation instructions.

#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.org/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Addressed currently open issue (replace XXXXX with an issue no in next line)

Closes #16642

### CODE REVIEW

File: sample.docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Docker Compose sample file for FreeCodeCamp
```
Comment (raisedadead) [inline sample.docker-compose.yml]: `FreeCodeCamp` => `freeCodeCamp`

File: sample.docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    image: node:8.9.4
+    volumes:
+      - .:/app
+    working_dir: /app
+    depends_on:
+      - mongo
+    ports:
+      - "3000:3000"
+      - "3002:3002"
+      - "8000:8000"
```
Comment (raisedadead) [inline sample.docker-compose.yml]: Which process is running on 8000?

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -288,6 +288,28 @@ node seed
...
+### Setup freeCodeCamp via Docker
+
+Create docker-compose.yml from sample file
+```bash
+cp sample.docker-compose.yml docker-compose.yml
+```
+Open .env and rename host from `localhost` to `mongo`
+
+Setup and run project:
+
```
Comment (raisedadead) [inline CONTRIBUTING.md]: Remove extra line

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -288,6 +288,28 @@ node seed
 npm run develop
 ```
 
+### Setup freeCodeCamp via Docker
+
+Create docker-compose.yml from sample file
+```bash
+cp sample.docker-compose.yml docker-compose.yml
+```
+Open .env and rename host from `localhost` to `mongo`
```
Comment (raisedadead) [inline CONTRIBUTING.md]: Add explicit instructions on what is to be changed

`change MONGOHQ_URL='mongodb://localhost:27017/freecodecamp' to MONGOHQ_URL='mongodb://mongo:27017/freecodecamp'`

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 4 above)
Comment (Bouncey) [reply CONTRIBUTING.md]: would this not be better handled in the `compose` file under `environment`?

Comment (raisedadead) [reply CONTRIBUTING.md]: Yup, that should be OK.

**Comment by otarza** [comment]
@raisedadead @Bouncey thanks for the review, I have updated PR according your requests.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
So far this looks good. Maybe we could use a link to the docker install instructions? I have tested only on macOS so far will wait for other platforms.

**Comment by otarza** [comment]
@raisedadead you mean link to this section? https://github.com/otarza/freeCodeCamp/blob/feature/16642-redesign-docker-env/CONTRIBUTING.md#setup-freecodecamp-via-docker

I have tested this on linux - CentOS 7, it should work on any platform with Docker installed, though I have not tested on Windows.

**Comment by otarza** [comment]
Ah, you mean installation instructions to Docker and docker-compose, yes, that would be good, I'll add those.

**Comment by tchaffee** [comment]
I tested on Ubuntu 16.04 and docker seems to work ok, but I'm getting an app error when I click on the 'Map' menu item in the browser:

```
GET 52.727 ms - /services/map?_csrf=xxx&returnMeta=trueis
error HttpError: Invalid Fetchr Access
```

I ran the following to bring the app up in Docker:

```
docker-compose run --rm server npm install
docker-compose run --rm server npm run only-once
docker-compose up
```

**Comment by otarza** [comment]
Hi @tchaffee, 

Yes, docker-compose commands you have used are correct.

I have just tested on Ubuntu 16.04, clean install with docker only. 

I was not able to reproduce same issue, it might not be docker related?

**Comment by otarza** [comment]
@raisedadead I have added docker and docker-compose links. 

Also I'm not sure why is snyk failing and I cant see the details as I don't have access to it.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -288,6 +288,27 @@ node seed
...
+# Setup project
+docker-compose run --rm node npm install
+docker-compose run --rm node npm run only-once
+```
+```bash
+# Run project
+docker-compose up
+```
+
+
```
Comment (raisedadead) [inline CONTRIBUTING.md]: nit: remove extra new line

File: sample.docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+#   docker-compose run --rm node npm install
+#   docker-compose run --rm node npm run only-once
+#
+# Run with:
+#   docker-compose up
+#
+version: "3"
+services:
+  node:
+    image: node:8.9.4
```
Comment (raisedadead) [inline sample.docker-compose.yml]: lets just make this `node:carbon` to be on latest LTS

File: sample.docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      - mongo
+    ports:
+      - "3000:3000"
+      - "3002:3002"
+    environment:
+      - MONGOHQ_URL=mongodb://mongo:27017/freecodecamp
+    command: npm run develop
+
+  mongo:
+    image: mongo:3.0.15
```
Comment (raisedadead) [inline sample.docker-compose.yml]: Can this be done to latest LTS with a named tag too?

**Comment by raisedadead** [comment]
Hi @otarza some real nice work here. We would love to have this as soon as we can, to improve our DX.

Some additional comments, since you are the expert:

1. While I understand that Dockerfile ( as a `Dockerfile.development` ) not needed, because we are not using images, but I think its worth it for customizing this in future (also, useful for automating when we do move to docker for deployments on production, whenever we plan to do so, in future)

So do you think, you can add that?

2. I found some sporadic  app crashes with nodemon. By any chance can we automate its restart? 
This can be ignored for now, if its too much work for you at this moment.

3. Windows doesn't play well in my QA (I know, its always a pain on Windows, but that's something that we so have to consider, as the goal of this to make contributing seemless).

Finally,
Would you need any inputs from our side to know about the workflow of editing challenges, so that things like, doing the intial seeding, installing dependencies, and seeding the database when files are edited, etc can be added as commands?

**Comment by raisedadead** [comment]
Also, if we can it would worth to suppress logs that we do not absolutely need for instance those of mongodb and mailhog services

Here is an example https://github.com/paralect/docker-compose-starter/blob/master/docker-compose.yml

Thanks a lot once again for taking the time in preparing this. It would help us a lot.

**Comment by ojongerius** [comment]
Hey there, great work so far 😺  

I'm sceptical of the strategy of having a sample file, and adding the Docker Compose file to .gitignore.
What were the use cases you had in mind that made you chose this solution?
Could those cases be solved by using an override file (`docker-compose.override.yml`)? If so I'd suggest keeping a default `docker-compose.yml`, and adding `docker-compose-override.yml` to `.gitignore`.

Thoughts?

**Review by ojongerius — CHANGES_REQUESTED** [review]
I suggest using override files instead of moving `docker-compose.yml` to `sample.docker-compose.yml`. Then adding `docker-compose.override.yml` to `.gitignore`.

Similar effect, better DX.

**Comment by ojongerius** [comment]
How is coming along @otarza?  I'd like to make some changes using over ride files, and don't want to interfere with your work.

**Comment by raisedadead** [comment]
@ojongerius I can see the no of issues we have been made aware of in the chat, and I am all up for improving the DX. 

And open-api sure has some dependencies on this repo's packaging /cc @QuincyLarson 

So, while this PR is not ready, I think we can make an execption, merge it as it is. And probably you could make another PR or update your existing ones to address those?

**Comment by raisedadead** [comment]
Hi @otarza thanks a lot for your contribution. We understand that you have done a nice work in preparing this, I have merged the changes.

@ojongerius could you please take a look at the review comments on this and make / update the PR for the additional needs?

**Pull request merged**